sys/kern/vm_map.c

```diff
@@ -340,17 +340,35 @@ vm_map_t *vm_map_clone(vm_map_t *map) {
   WITH_MTX_LOCK (&map->mtx) {
     vm_segment_t *it;
     TAILQ_FOREACH (it, &map->entries, link) {
+      vm_seg_flags_t flags = it->flags;
       vm_object_t *obj;
       vm_segment_t *seg;
+      vm_object_t *shadow;
 
       if (it->flags & VM_SEG_SHARED) {
         refcnt_acquire(&it->object->ref_counter);
         obj = it->object;
       } else {
-        /* vm_object_clone will clone the data from the vm_object_t
-         * and will return the new object with ref_counter equal to one */
-        obj = vm_object_clone(it->object);
+        if ((it->flags & VM_SEG_NEED_COPY) != 0 && it->object->npages == 0) {
+          obj = vm_object_alloc(VM_SHADOW);
+          shadow = it->object->shadow_object;
+          obj->shadow_object = shadow;
+        } else {
+          shadow = it->object;
+          obj = vm_object_alloc(VM_SHADOW);
+          obj->shadow_object = shadow;
+          it->object = vm_object_alloc(VM_SHADOW);
+          it->object->shadow_object = shadow;
+          vm_object_set_readonly(shadow);
```

Comment:
Could you explain what should be done here with pages marked as `VM_PROT_READ | VM_PROT_EXEC`?

Reply:
I think in case when sth is executable, then we also should set is as readonly and just disregard the fact that this is VM_PROT_EXEC, don't do anything special in that case

Reply:
Maybe I don't understand something but if you mark page as RO you can't execute any instruction from that page so how it work's in that case? Do you kill new process in that case or make copy of .text?